### PR TITLE
feat: Add Advanced Filtering Options to Invoice Dashboard and Improve Balance Check

### LINKED ISSUE #183: [Request Invoicing] Improve filtering on Invoice Dashboard

### PR DESCRIPTION
**Pull request by sstefdev**
Fixes #183 

### Problem
Users requested the ability to hide voided or canceled requests from the Invoice Dashboard and introduced filters for better usability:

- Type filter (IN/OUT) for the All tab.
- Status filter for easier navigation of requests.
- Chain filter (dependent on the payment chain column from #70).

### Changes:

- Filters:
   - Filters work on all of the tabs (All, Pay, Get Paid)
   - Added advanced filters (Type, Status, and Chain) to the All tab.
- Balance Check:
   - Improved balance check logic with better error messages and detailed feedback.
- New Components:
   - Created a searchable-dropdown-checkbox component to support multiple filters.
   - Updated styles and usability for dropdown components.
- Dashboard Logic:
   - Refactored filtering logic for scalability and maintainability.

https://github.com/user-attachments/assets/3644f815-9cdc-4fce-95d3-d803c0d17d2b

## Summary by CodeRabbit

- **New Features**
	- Enhanced error handling and logic for payment processing in the invoice view.
	- Improved filtering capabilities for requests in the dashboard with a new searchable dropdown.
	- Introduced a new searchable checkbox dropdown component for multi-option selection.

- **Bug Fixes**
	- Improved error handling and user feedback during request loading and payment detection.

- **Style**
	- Updated layouts and styles for dropdowns and input components to enhance usability.

### CODE REVIEW

File: shared/components/dropdown.svelte [hunk 1]
```diff
@@ -13,10 +19,16 @@
   let dropdownId: string;
   let isOpen = false;
   let dropdownContainer: HTMLElement;
+  let localOptions = options;
```
[the remark this answers is not in the document]
Comment (sstefdev) [reply shared/components/dropdown.svelte:22]: It is already reactive without this

File: shared/components/dropdown.svelte [hunk 2]
```diff
@@ -77,30 +89,22 @@
...
+          {#each localOptions as option}
+            <li
+              class="dropdown-item"
+              class:selected={option.checked}
+              on:click={() => selectOption(option.value, option.checked)}
+            >
+              <span>{option.label}</span>
+              <div class="custom-checkbox" class:checked={option.checked}>
+                {@html option.checked ? CheckIcon : ""}
+              </div>
```
[the remark this answers is not in the document]
Comment (sstefdev) [reply shared/components/dropdown.svelte:101]: Yes I will add this

File: shared/components/dropdown.svelte [hunk 3]
```diff
@@ -179,61 +173,41 @@
...
+    display: flex;
+    align-items: center;
+    justify-content: center;
   }
 
-  .dropdown-checkbox-option:hover {
-    background-color: #f3f4f6;
+  .custom-checkbox.checked {
+    background-color: var(--mainColor);
+    border-color: var(--mainColor);
```
[the remark this answers is not in the document]
Comment (sstefdev) [reply shared/components/dropdown.svelte:211]: Will add this as well

File: packages/invoice-dashboard/src/lib/view-requests.svelte [hunk 4]
```diff
@@ -222,13 +260,29 @@
...
-      (currentTab === "Pay" &&
-        request.payer?.value?.toLowerCase() === signer?.toLowerCase())
+      networkMatch &&
+      txTypeMatch &&
+      statusMatch &&
+      (currentTab === "All" ||
+        (currentTab === "Get Paid" &&
+          request.payee?.value?.toLowerCase() === signer?.toLowerCase()) ||
+        (currentTab === "Pay" &&
+          request.payer?.value?.toLowerCase() === signer?.toLowerCase()))
```
[the remark this answers is not in the document]
Comment (sstefdev) [reply packages/invoice-dashboard/src/lib/view-requests.svelte:285]: No tests for now

File: shared/components/searchable-checkbox-dropdown.svelte [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                <NetworkIcon network={option.value} />
+              </div>
+            {:else if type === "transaction"}
+              <TxType type={option.value} />
+            {:else if type === "status"}
+              <StatusLabel status={option.value} />
+            {/if}
+            <div class="custom-checkbox" class:checked={option.checked}>
+              {@html option.checked ? CheckIcon : ""}
+            </div>
```
[the remark this answers is not in the document]
Comment (sstefdev) [reply shared/components/searchable-checkbox-dropdown.svelte:163]: It is the same thing at the end of the day, an XSS can not come out of this

File: shared/components/searchable-checkbox-dropdown.svelte [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+          </li>
+        {/each}
+      </ul>
+    </div>
+  {/if}
```
[the remark this answers is not in the document]
Comment (sstefdev) [reply shared/components/searchable-checkbox-dropdown.svelte:168]: No need for this

**Comment by MantisClone** [comment]
Inconsistent placeholder text: "Filter by Status" vs "Type"

Instead of "Filter By", can you add a Filter icon for all three filters?

**Comment by MantisClone** [comment]
Does the Type filter appear only on the All tab? Or does it appear on the Pay and Get Paid tabs too?

The Type filter must not conflict with the Pay and Get Paid tabs. How we achieve this is debatable. Before I suggest possible solutions I want to know if you've already thought of this and resolved it in some way.

**Comment by sstefdev** [comment]
@MantisClone 

1. We don't have any icons on the design that Mia added, what filter icons should I add?
2. Why would the tabs get in the way of filtering? The logic seperate.

**Comment by MantisClone** [comment]
1. Fair point. I still think we should remove "Filter by" from the placeholder text because it makes the filter boxes too wide.
2. The pay and get paid tabs are effectively the same as the in and out filters. Imagine the user is on the pay tab and filters so only "IN" payments are shown. This is an impossible combination so zero requests would be shown.

**Review by MantisClone — COMMENTED** [review]
Partial review

File: packages/invoice-dashboard/src/lib/dashboard/invoice-view.svelte [hunk 7]
```diff
@@ -372,24 +372,78 @@
...
-          signer,
-          requestData.expectedAmount
-        ),
-    };
+    try {
+      if (!paymentNetworkExtension?.id || !address || !signer) {
+        console.log("Missing dependencies:", {
+          network: paymentNetworkExtension?.id,
+          address,
+          signer: !!signer,
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/invoice-dashboard/src/lib/dashboard/invoice-view.svelte:380]: Nevermind. I figured it out. According to the AI:
>In JavaScript, the `!!` (double negation) prefix is not a specific operator. Instead, it's a combination of two negation operators (`!`) used to convert a value to a boolean (`true` or `false`).
Here's how it works:
> * The first `!` converts the value to a boolean, inverting its truthiness.
> * The second `!` inverts the result again, giving you the original value's truthiness as a boolean.

So this expression looks OK to me. Resolving as No Problem :ok_person:

File: packages/invoice-dashboard/src/lib/dashboard/invoice-view.svelte [hunk 8]
```diff
@@ -373,24 +373,68 @@
...
+      const expectedChainId = getNetworkIdFromNetworkName(network);
+      const expectedChainIdNumber = parseInt(expectedChainId, 16);
+
+      if (chainId !== expectedChainIdNumber) {
+        console.error("Wrong network:", {
+          current: `0x${chainId.toString(16)}`,
+          expected: expectedChainId,
+        });
+        return false;
+      }
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/invoice-dashboard/src/lib/dashboard/invoice-view.svelte:409]: As long as we're using `!==` and not `!=` - meaning that it compares both value and type - then I think we're okay. Resolving as No Problem :ok_person:

**Comment by MantisClone** [comment]
@coderabbitai full review

**Pull request merged**